### PR TITLE
add parseMergeResult functionality and tests to detect merge conflicts 

### PR DESCRIPTION
**Pull request by shiftkey**
This PR is the underlying Git work to support #4588. I generated a bunch of test data by scanning the `desktop/desktop`, `electron/electron` and `microsoft/vscode` repositories for examples of merge conflicts - these are now baked as tests into the project, and caught some corner cases I hadn't uncovered with the original code.

For the moment `parseMergeResult`  just scans the Git output to build up an array of entries, to then determine what files have conflict. But there's already details in the diff output about _how_ files conflict that we can use down the track.

For reference, this is how I generated the merge conflict files - if there's interest in me adding this as something more reusable to regenerate tests cases when we upgrade `dugite` let me know and we can figure out how to do that:

```ts
  const repo = new Repository('/Users/shiftkey/src/vscode', -1, null, false)
  const allBranches = await getBranches(repo)
  const start =
    allBranches.find(b => b.nameWithoutRemote === 'master') || null

  if (start == null) {
    throw new Error('Unable to find current branch!')
  }

  await checkoutBranch(repo, null, start)

  const status = await getStatusOrThrow(repo)

  if (status.currentBranch == null) {
    throw new Error('Unable to find current branch!')
  }

  const ours =
    allBranches.find(b => b.nameWithoutRemote === status.currentBranch) ||
    null

  if (ours == null) {
    throw new Error('Unable to find ours!')
  }

  const otherBranchesToCompare = allBranches.filter(
    b => b.tip.sha !== ours.tip.sha
  )

  for (const theirs of otherBranchesToCompare) {
    const fileName = `merge-${theirs.nameWithoutRemote}-into-${
      ours.name
    }.txt`.replace('/', '-')

    const mergeBase = await getMergeBase(repo, ours.tip.sha, theirs.tip.sha)

    if (mergeBase === ours.tip.sha) {
      continue
    }

    if (mergeBase === theirs.tip.sha) {
      continue
    }

    const result = await GitProcess.exec(
      ['merge-tree', mergeBase, ours.tip.sha, theirs.tip.sha],
      repo.path,
      { maxBuffer: 100 * 1024 * 1024 }
    )
    const stdout = result.stdout

    if (stdout.indexOf('<<<<<<') === -1) {
      continue
    }

    const relativePath = Path.join(
      __dirname,
      `../../fixtures/merge-parser/vscode/${fileName}`
    )

    await FSE.writeFile(relativePath, stdout)
  }
})
```

### CODE REVIEW

File: app/src/lib/git/merge.ts [hunk 1]
```diff
@@ -21,3 +25,42 @@ export async function getMergeBase(
...
+ * @param repository The repository containing the branches to merge
+ * @param ours The current branch
+ * @param theirs Another branch to merge into the current branch
+ */
+export async function mergeTree(
+  repository: Repository,
+  ours: Branch,
+  theirs: Branch
+): Promise<MergeResult | null> {
+  console.time('getMergeBase')
```
Comment (shiftkey) [inline app/src/lib/git/merge.ts]: I'm going to clean these up after doing the UI work, but for now these are nice to help identify chronic cases with any of the operations underneath.

File: app/src/lib/git/merge.ts [hunk 2]
```diff
@@ -9,15 +13,73 @@ export async function merge(
   await git(['merge', branch], repository.path, 'merge')
 }
 
+/**
+ * Find the base commit between two refs
+ *
+ * @returns the commit id of the merge base, or null if the two refs do not
```
Comment (iAmWillShepherd) [inline app/src/lib/git/merge.ts]: We should document what type of string we're looking for for the _*Ref_ parameters (like is it expecting a name of a branch or tag, a commit, or some other value).

File: app/src/lib/git/merge.ts [hunk 3]
```diff
@@ -9,15 +13,73 @@ export async function merge(
...
   secondRef: string
-): Promise<string> {
+): Promise<string | null> {
   const process = await git(
     ['merge-base', firstRef, secondRef],
     repository.path,
-    'merge-base'
+    'merge-base',
+    {
+      successExitCodes: new Set([0, 1]),
```
Comment (iAmWillShepherd) [inline app/src/lib/git/merge.ts]: What do these exit codes mean? Is it possible to pull these codes into constant so we don't have to throw around magic numbers?

File: app/src/lib/git/merge.ts [hunk 4]
```diff
@@ -9,15 +13,73 @@ export async function merge(
...
+  ours: Branch,
+  theirs: Branch
+): Promise<MergeResult | null> {
+  console.time('getMergeBase')
+  const mergeBase = await getMergeBase(repository, ours.tip.sha, theirs.tip.sha)
+  console.timeEnd('getMergeBase')
+
+  if (mergeBase === null) {
+    // TODO: return a better result to indicate something is wrong
+    return null
```
Comment (iAmWillShepherd) [inline app/src/lib/git/merge.ts]: Could we throw an error here or will you change this as a part of some other work you have planned?

File: app/src/lib/git/merge.ts [hunk 5]
```diff
@@ -9,15 +13,73 @@ export async function merge(
...
+  }
+
+  if (mergeBase === ours.tip.sha) {
+    return { kind: MergeResultKind.Success, entries: [] }
```
Comment (iAmWillShepherd) [inline app/src/lib/git/merge.ts]: This can be pulled out into a variable since it's being used below.

File: app/src/lib/git/merge.ts (hunk 2 above)
Comment (shiftkey) [reply app/src/lib/git/merge.ts]: The docs for `merge-base` say `<commit>` but my testing indicates this can be any commit/ref/tag/`HEAD`. I'll make it clearer in the title (documenting each `@param` felt noisy at the time):

> Find the base commit between two refs

File: app/src/lib/merge-tree-parser.ts [hunk 6]
```diff
@@ -0,0 +1,194 @@
+import { IMergeEntry, MergeResult, MergeResultKind } from '../models/merge'
+
+function updateCurrentMergeEntry(
+  entry: IMergeEntry | undefined,
+  context: string,
+  blobSource: {
```
Comment (iAmWillShepherd) [inline app/src/lib/merge-tree-parser.ts]: This should be pulled out into it's own type.

File: app/src/lib/git/merge.ts (hunk 4 above)
Comment (shiftkey) [reply app/src/lib/git/merge.ts]: This is handled in #5495 because of a corner case I wrote more about here: https://github.com/desktop/desktop/issues/4588#issuecomment-417382948

```
  if (mergeBase === null) {
    return { kind: MergeResultKind.Invalid }
  }
```

File: app/src/lib/git/merge.ts (hunk 3 above)
Comment (shiftkey) [reply app/src/lib/git/merge.ts]: The docs aren't really clear on what exit codes are expected here but from my testing showed `0` means a common ancestor was found. `1` means one could not be found (no `stdout` or `stderr` occurred).

Given the rest of the function is rather boring does introducing constants here help readability, or is there something else we can do here?

Comment (iAmWillShepherd) [reply app/src/lib/git/merge.ts]: I think it's a readability improvement. If we can pass the exit codes we care about to the `git` function, I think we should always use a name instead of numbers. I've seen this in a few places now, so I figured passing these codes will become more common in the future.

File: app/src/lib/git/merge.ts (hunk 2 above)
Comment (iAmWillShepherd) [reply app/src/lib/git/merge.ts]: Yea, changing the name of the param to `committish` may also work. I agree that having the param docs is a bit noisey.a

File: app/src/lib/git/merge.ts (hunk 3 above)
Comment (shiftkey) [reply app/src/lib/git/merge.ts]: >  I've seen this in a few places now, so I figured passing these codes will become more common in the future.

Are we talking about `expectedErrors` (where an error message matches a known regex, and we map this to an enum in code) rather than `successExitCodes` (the raw numbers)?

I'm definitely in favour of preferring `expectedErrors` where it makes sense, but we don't get any `stdout` or `stderr` in this case to indicate anything went wrong - just the `1` exit code. [Here's the relevant Git source code, for reference](https://github.com/git/git/blob/2f743933341f276111103550fbf383a34dfcfd38/builtin/merge-base.c#L11-L28).

File: app/src/lib/merge-tree-parser.ts [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+const blobEntryRe = /^\s{2}(result|our|their|base)\s+(\d{6})\s([0-9a-f]{40})\s(.+)$/
+
+/**
+ * Parse the Git output of a merge-tree command to identify whether it
+ * has detected any conflicts between the branches to be merged
+ *
+ * @param text the stdout from a `git merge-tree` command
+ *
+ */
+export function parseMergeResult(text: string): MergeResult {
```
Comment (iAmWillShepherd) [inline app/src/lib/merge-tree-parser.ts]: This seems like a good thing to add some tests for.

Comment (shiftkey) [reply app/src/lib/merge-tree-parser.ts]: Check out `app/test/unit/git/parse-merge-result-test.ts`

Comment (iAmWillShepherd) [reply app/src/lib/merge-tree-parser.ts]: ahh, it's at the bottom and the diff is collapsed.
